### PR TITLE
[TGL] KWD: performance improvements

### PR DESCRIPTION
**Pull request by mrajwa**
This set of patches aims to improve KWD overal performance targeting TGL platforms.

Signed-off-by: Marcin Rajwa <marcin.rajwa@linux.intel.com>

### CODE REVIEW

**Comment by cujomalainey** [comment]
@bzhg FYI

File: tools/topology/sof-tgl-max98357a-rt5682.m4 [hunk 1]
```diff
@@ -84,7 +84,7 @@ define(DMIC_PIPELINE_16k_ID, `11')
 define(DMIC_PIPELINE_KWD_ID, `12')
 define(DMIC_DAI_LINK_16k_ID, `2')
 # define pcm, pipeline and dai id
-define(KWD_PIPE_SCH_DEADLINE_US, 20000)
+define(KWD_PIPE_SCH_DEADLINE_US, 5000)
```
Comment (plbossart) [inline tools/topology/sof-tgl-max98357a-rt5682.m4:87]: why does this improve performance and what does 'performance' means here.

File: src/audio/kpb.c [hunk 2]
```diff
@@ -1244,8 +1245,11 @@ static enum task_state kpb_draining_task(void *arg)
 		 */
 			comp_cl_info(&comp_kpb, "kpb: update history_depth by %d",
 				     *rt_stream_update);
+			spin_lock_irq(&kpb->lock, flags);
 			history_depth += *rt_stream_update;
 			*rt_stream_update = 0;
+			spin_unlock_irq(&kpb->lock, flags);
+
```
Comment (plbossart) [inline src/audio/kpb.c:1252]: the only case where the kpb->lock is used is for the transition to the INIT_DRAINING state
here you are already draining, so how does the lock help?

File: src/audio/kpb.c [hunk 3]
```diff
@@ -1252,6 +1252,15 @@ static enum task_state kpb_draining_task(void *arg)
...
 	}
+
+	while (!sink->stream.free) {
+		/* This is to workaround HW problem of no DMA complete
+		 * IRQ. Because of that we have to be be sure HOST DMA
+		 * has some free space otherwise real time copy may be
+		 * stuck.
+		 */
+		comp_copy(sink->sink);
+	}
```
Comment (plbossart) [inline src/audio/kpb.c:1263]: This again looks like a loop instead of something based on timers and/or signaling. Not so good.

File: src/audio/detect_test.c [hunk 4]
```diff
@@ -121,9 +120,9 @@ static void notify_kpb(const struct comp_dev *dev)
 	cd->client_data.sink = NULL;
 	cd->client_data.id = 0; /**< TODO: acquire proper id from kpb */
 	/* time in milliseconds */
-	cd->client_data.history_depth = (cd->history_depth != 0) ?
-					 cd->history_depth :
-					 cd->config.history_depth;
+	cd->client_data.drain_req = (cd->drain_req != 0) ?
+					 cd->drain_req :
+					 cd->config.drain_req;
```
Comment (plbossart) [inline src/audio/detect_test.c:125]: I don't see how drain_req is clearer than history_depth. No idea what drain_req means and what the concept is here.

File: src/platform/tigerlake/include/platform/lib/memory.h [hunk 5]
```diff
@@ -261,7 +261,7 @@
 	HEAP_RT_COUNT1024 * 1024 + HEAP_RT_COUNT2048 * 2048 + \
 	HEAP_RT_COUNT4096 * 4096)
 
-#define HEAP_BUFFER_SIZE	0x50000
+#define HEAP_BUFFER_SIZE	0x80000
```
Comment (plbossart) [inline src/platform/tigerlake/include/platform/lib/memory.h:264]: why is this needed? to compensate for the resume latency?
That looks like a very large step in terms of allocation.

File: tools/topology/sof-tgl-max98357a-rt5682.m4 (hunk 1 above)
Comment (mrajwa) [reply tools/topology/sof-tgl-max98357a-rt5682.m4:87]: @plbossart here by *performace* improvement I mean how much work we can do in 1 [ms] time period. This patch improves performance because we process less samples (4 times less) and therefore lowered CPU load.

File: src/audio/kpb.c (hunk 2 above)
Comment (mrajwa) [reply src/audio/kpb.c:1252]: It disables interrupts which are a concern here because if they happen in between 'history_depth += *rt_stream_update;' and '*rt_stream_update = 0;' you lose the real time stream which happened when we were draining HB.

File: src/audio/detect_test.c (hunk 4 above)
Comment (mrajwa) [reply src/audio/detect_test.c:125]: You have no idea what `drain_req` is? What about "draining request"? When working with validation teams I notice they are confused by "history depth" and somehow "draining request" is more meaningful to people.

File: src/platform/tigerlake/include/platform/lib/memory.h (hunk 5 above)
Comment (mrajwa) [reply src/platform/tigerlake/include/platform/lib/memory.h:264]: The commit messages says " Thanks to this change we can provide KPB more space for history data"

File: src/audio/kpb.c (hunk 3 above)
Comment (mrajwa) [reply src/audio/kpb.c]: Agree, I removed it and added better solution.

File: tools/topology/sof-tgl-max98357a-rt5682.m4 (hunk 1 above)
Comment (lgirdwood) [reply tools/topology/sof-tgl-max98357a-rt5682.m4:87]: but dont we still have to process the full draining buffer ? This just looks like we are splitting into smaller chunks, bust should EDF be pre-emptive anyway so chunk size should not matter here ?

Comment (mrajwa) [reply tools/topology/sof-tgl-max98357a-rt5682.m4:87]: @lgirdwood no, this has nothing to do with draining/history buffer. This is about the real time steam - when its size is 20 [ms] then you need more MIPS to process it than when the period is only 5 ms.

**Comment by mmaka1** [comment]
@mrajwa @slawblauciak The patches seem to fail the same set of tests (rebuild of Internal Intel shows consistent failures).

**Comment by slawblauciak** [comment]
Fixed conflicts.

**Comment by q009** [comment]
@lgirdwood tests clean, would much appreciate if it could get merged soon, as it's quite critical. Thoughts?

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@q009 @slawblauciak @mrajwa can you rebase. Think we are good to go.

**Comment by slawblauciak** [comment]
@lgirdwood what for? There are no conflicts, I fixed them earlier.

**Comment by mrajwa** [comment]
@lgirdwood @slawblauciak  rebased.

**Comment by lgirdwood** [comment]
Jenkins shows SDW codec kernel error and BYT BAT test sounds good.

**Comment by lgirdwood** [comment]
@mrajwa @zrombel it seems internal CI has been pending for a while now. Can you check and let me know if we are good to merge.

**Comment by zrombel** [comment]
Probably there was a queue in QuickBuild. At this point all tests are PASS and PR seems to be good to merge.

**Comment by lgirdwood** [comment]
CI showing unrelated SDW kernel errors on CML
